django/middleware/csrf.py

```diff
@@ -221,7 +222,41 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                 # branches that call reject().
                 return self._accept(request)
 
-            if request.is_secure():
+            # Verify the Origin header, if present.
+            origin_verified = False
+            if 'HTTP_ORIGIN' in request.META:
```

Comment:
Have you considered putting the logic beneath this `if` block into a separate, perhaps independently testable function?

Reply:
Mhm that is a good idea, this way we could actually unit test it nicely. I would not make it public though (ie we shouldn't document it)

Reply:
I guess I'm not completely opposed but IMO it looks a little funny to have some logic (referrer checking) inlined and this elsewhere. I'm not sure it eases testing much if the signature is something like `def _origin_verified(self, request)` considered a complete request object needs to be created. Alternatively, the relevant request values could be extract outside the method and the arguments could be something like `(good_origin, expected_origin)`   Something like:
```diff
diff --git a/django/middleware/csrf.py b/django/middleware/csrf.py
index a1e28862cb..9043ce41da 100644
--- a/django/middleware/csrf.py
+++ b/django/middleware/csrf.py
@@ -203,6 +203,37 @@ class CsrfViewMiddleware(MiddlewareMixin):
             # Use same token next time.
             request.META['CSRF_COOKIE'] = csrf_token
 
+    def _origin_verified(self, request):
+        good_origin = '%s://%s' % (
+            'https' if request.is_secure() else 'http',
+            request.get_host(),
+        )
+        allowed_origins = [good_origin]
+        # Tuples of (allowed scheme, allowed netloc), where all
+        # subdomains of netloc are allowed.
+        allowed_origin_subdomains = []
+        for trusted_origin in settings.CSRF_TRUSTED_ORIGINS:
+            if '*' in trusted_origin:
+                parsed = urlparse(trusted_origin)
+                allowed_origin_subdomains.append(
+                    (parsed.scheme, parsed.netloc.lstrip('*'))
+                )
+            else:
+                allowed_origins.append(trusted_origin)
+        # Reject the request if the origin header doesn't match an
+        # allowed value.
+        expected_origin = request.META['HTTP_ORIGIN']
+        if expected_origin in allowed_origins:
+            return True
+        parsed_origin = urlparse(expected_origin)
+        expected_scheme = parsed_origin.scheme
+        expected_netloc = parsed_origin.netloc
+        for allowed_scheme, allowed_host in allowed_origin_subdomains:
+            if (allowed_scheme == expected_scheme and
+                    is_same_domain(expected_netloc, allowed_host)):
+                return True
+        return False
+
     def process_view(self, request, callback, callback_args, callback_kwargs):
         if getattr(request, 'csrf_processing_done', False):
             return None
@@ -225,35 +256,8 @@ class CsrfViewMiddleware(MiddlewareMixin):
             # Verify the Origin header, if present.
             origin_verified = False
             if 'HTTP_ORIGIN' in request.META:
-                good_origin = '%s://%s' % (
-                    'https' if request.is_secure() else 'http',
-                    request.get_host(),
-                )
-                allowed_origins = [good_origin]
-                # Tuples of (allowed scheme, allowed netloc), where all
-                # subdomains of netloc are allowed.
-                allowed_origin_subdomains = []
-                for trusted_origin in settings.CSRF_TRUSTED_ORIGINS:
-                    if '*' in trusted_origin:
-                        parsed = urlparse(trusted_origin)
-                        allowed_origin_subdomains.append(
-                            (parsed.scheme, parsed.netloc.lstrip('*'))
-                        )
-                    else:
-                        allowed_origins.append(trusted_origin)
-                # Reject the request if the origin header doesn't match an
-                # allowed value.
-                expected_origin = request.META['HTTP_ORIGIN']
-                if expected_origin not in allowed_origins:
-                    parsed_origin = urlparse(expected_origin)
-                    expected_scheme = parsed_origin.scheme
-                    expected_netloc = parsed_origin.netloc
-                    for allowed_scheme, allowed_host in allowed_origin_subdomains:
-                        if (allowed_scheme == expected_scheme and
-                                is_same_domain(expected_netloc, allowed_host)):
-                            break
-                    else:
-                        return self._reject(request, REASON_BAD_ORIGIN % expected_origin)
+                if not self._origin_verified(request):
+                    return self._reject(request, REASON_BAD_ORIGIN % request.META['HTTP_ORIGIN'])
                 origin_verified = True
 
             if request.is_secure() and not origin_verified:
```
I'm not sure if additional tests will bring much benefit.

Reply:
Regarding the referrer checking, given the length and complexity of the containing function, I was thinking it would be good to have that in a separate function, too. But I didn't want to suggest changes outside the scope of your change. I think either signature you proposed would be an improvement. (Another option, in between, would be to pass `good_origin` and `request.META`.) Perhaps the decision can be guided by what the signatures would be for other portions of the logic if they were similarly broken out (for parallel structure).

Reply:
What about changing the signature to `_origin_verified(expected_origin, request_origin)` -- this way we'd decouple it from the request and would allow for easy testing. Granted the naming of the params might need a little bit of thinking but you get the idea :)

Reply:
Please check the last commit I just pushed. Are you insisting on additional tests? I think all cases are covered but I will double-check.

Reply:
I haven't looked at the tests to closely yet, but I think you might have covered them all. So no I am not insisting on additional tests. I suggested the signature change so we get a method with smaller scope that can get tested more easily (ie two params instead of having to go through the request creation dance). But I have no strong feeling either way.